libtenzir/builtins/aggregation-functions/max.cpp

```diff
@@ -42,7 +49,95 @@ class max_function final : public aggregation_function {
   std::optional<type_to_data_t<Type>> max_ = {};
 };
 
-class plugin : public virtual aggregation_function_plugin {
+// TODO: Probably merge this with min
+class max_instance final : public aggregation_instance {
+public:
+  using max_t = variant<caf::none_t, int64_t, uint64_t, double, duration>;
+  explicit max_instance(ast::expression expr) : expr_{std::move(expr)} {
+  }
+
+  auto update(const table_slice& input, session ctx) -> void override {
+    if (max_ and std::holds_alternative<caf::none_t>(max_.value())) {
+      return;
+    }
+    auto arg = eval(expr_, input, ctx);
+    if (not type_) {
+      type_ = arg.type;
+    }
+    const auto warn = [&](const auto&) -> max_t {
+      diagnostic::warning("expected `{}`, got `{}`", type_, arg.type)
+        .primary(expr_)
+        .emit(ctx);
+      return caf::none_t{};
+    };
+    auto f = detail::overload{
+      [](const arrow::NullArray&) {},
+      [&]<class T>(const T& array)
+        requires numeric_type<type_from_arrow_t<T>>
+      {
+        for (auto i = int64_t{}; i < array.length(); ++i) {
+          if (array.IsValid(i)) {
+            const auto val = array.Value(i);
+            if (not max_) {
+              max_ = val;
+              continue;
+            }
+            max_ = max_->match(
+              warn,
```

Comment:
So if `max_` has an unexpected type here, we set it to `caf::none` and continue with the next iteration. But we don't handle that then correctly in the next iteration, or do we?

Reply:
Since the type is set to `caf::none` on the next iteration it fails to match any other type and probably emits the warning again which is deduplicated. The output is as expected, but we could short-circuit here.